radicle/src/cob/identity.rs

```diff
@@ -0,0 +1,680 @@
+    }
+
+    pub fn published(&self) -> Option<Published> {
+        if self.revisions.is_closed() {
+            Some(Published {
+                title: self.title().to_string(),
+                description: self.description().map(|d| d.to_string()),
+            })
+        } else {
+            None
+        }
+    }
+
+    /// The most recent title for the proposal.
+    pub fn title(&self) -> &str {
+        self.title.get().get()
+    }
+
+    /// The most recent description for the proposal, if present.
+    pub fn description(&self) -> Option<&str> {
+        Some(self.description.get().get())
+    }
+
+    /// A specific [`Revision`], that may be redacted.
+    pub fn revision(&self, revision: &RevisionId) -> Gate<Option<&Redactable<Revision>>> {
+        self.revisions
+            .as_ref()
+            .map(|revisions| revisions.get(revision))
+    }
+
+    /// All the [`Revision`]s that have not been redacted.
+    pub fn revisions(&self) -> impl DoubleEndedIterator<Item = (&RevisionId, &Revision)> {
+        self.revisions.iter().flat_map(|rs| rs.iter()).filter_map(
+            |(rid, r)| -> Option<(&RevisionId, &Revision)> { r.get().map(|r| (rid, r)) },
+        )
+    }
+
+    pub fn latest_by(&self, who: &PublicKey) -> Option<(&RevisionId, &Revision)> {
+        self.revisions().rev().find_map(|(rid, r)| {
+            if r.author.id() == who {
+                Some((rid, r))
+            } else {
+                None
+            }
+        })
+    }
+
+    pub fn latest(&self) -> Option<(&RevisionId, &Revision)> {
+        self.revisions().next_back()
+    }
+}
+
+impl store::FromHistory for Proposal {
+    type Action = Action;
+    type Error = ApplyError;
+
+    fn type_name() -> &'static TypeName {
+        &*TYPENAME
+    }
+
+    fn apply(&mut self, ops: impl IntoIterator<Item = Op>) -> Result<(), Self::Error> {
+        for op in ops {
+            let id = op.id();
+            let author = Author::new(op.author);
+            let timestamp = op.timestamp;
+
+            match op.action {
+                Action::Accept {
+                    revision,
+                    signature,
+                } => {
+                    let revisions = self.revisions.get_mut().ok_or(ApplyError::Published)?;
+
+                    match revisions.get_mut(&revision) {
+                        Some(Redactable::Present(revision)) => {
+                            revision.accept(op.author, signature, op.clock)
+                        }
+                        Some(Redactable::Redacted) => return Err(ApplyError::Redacted(revision)),
+                        None => return Err(ApplyError::Missing(revision)),
+                    }
+                }
+                Action::Edit { title, description } => {
+                    self.title.set(title, op.clock);
+                    self.description.set(description, op.clock);
+                }
+                Action::Publish => self.revisions.merge(Gate::closed()),
```

Comment:
Ok, I think I understand what you're going for here, but I'm not sure it's the right approach :thinking: 

For one, I think I'd adopt the same design as patch/issue, where there is a state that can be "open" or "closed". It avoids needing this gate type, and doesn't overwrite the data, ie. it can still be displayed if necessary.

Reply:
What you can do if you want to always merge towards a final state (least upper bound) is simply use a `Max<State>`, where the `State` enum's partial-ord is defined such that the greatest element is the "closed" state.

Reply:
Ya, I suppose that could work :) I guess I was trying to be too clever by stoping any publishing happening after the initial publish, but that can just be a check on the state.